Add a merge option to match.filter_by

olipovch opened the pull request:
### What does this PR do?
Adds a merge option to match.filter_by for merging a dictionary specified by the **merge** argument with the dictionary specified by the **lookup** argument.

### What issues does this PR fix or reference?
#49845

### Previous Behavior
No merge option existed in match.filter_by.

### New Behavior
Allows merging of dictionaries in match.filter_by.

### Tests written?

Yes; Unit tests created in test_match.py.

### Commits signed with GPG?

Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

olipovch committed:
Add a merge option to match.filter_by

Feature added for #49845

dwoz commented:
@olipovch Looks like there are a few unit test failures related to these changes, can you look into them?

olipovch commented:
Working on making the unit tests compatible with Python 3.

olipovch committed:
Add unit tests for match.filter_by

Create unit tests for the filter_by function of the match module.
Included are basic tests for filter_by functionality since previous
tests weren't defined as well as tests for the merge option.

olipovch commented:
The Jenkins tests passed initially but are now failing. Should the commits be rebased?

dwoz commented:
@olipovch Thanks, I've merge the latest changes into this branch. If the tests pass now we'll get this merged.

gtmanfred commented on salt/modules/match.py:
```diff
@@ -348,6 +351,16 @@ def filter_by(lookup,
+                if lookup[key] is None:
+                    return merge
+                else:
+                    salt.utils.dictupdate.update(lookup[key], copy.deepcopy(merge))
```
It would be nice to also be able to set merge_list=True on this.

olipovch replied:
Added the `merge_lists=True` option to the dictupdate.update call.

gtmanfred replied:
No, i think it should be configurable, it shouldn't always be on by default.

olipovch replied:
Updated `merge_lists` to be a configurable option in `match.filter_by`.

gtmanfred replied:
Awesome, thanks!

olipovch committed:
Add a merge_lists option to match.filter_by

Add a merge_lists option to the dictupdate.update call in
match.filter_by for specifying merge behavior of list values.

olipovch committed:
Add unit tests for merging lists in match.filter_by

Add unit tests when merging list values with the merge_lists=True option
and with the merge_lists=False option.
